src/DotNetty.Common/ThreadLocalPool.cs

```diff
@@ -12,55 +12,375 @@ public class ThreadLocalPool
+                        }
+                        element.Stack = dst;
+                        dst.Push(element);
+                        srcElems[i] = null;
+                    }
+
+                    if (srcEnd == LinkCapacity && head.next != null)
+                    {
+                        this.head = head.next;
+                    }
+
+                    head.ReadIndex = srcEnd;
+                    return true;
+                }
+                else
+                {
+                    // The destination stack is full already.
+                    return false;
+                }
             }
         }
 
-        internal sealed class Stack : Stack<Handle>
+        internal sealed class Stack
         {
-            public readonly ThreadLocalPool Owner;
+            public readonly ThreadLocalPool Parent;
             public readonly Thread Thread;
 
-            public Stack(int initialCapacity, ThreadLocalPool owner, Thread thread)
-                : base(initialCapacity)
+            internal Handle[] elements;
+
+            private int maxCapacity;
+            internal int head;
+
+            WeakOrderQueue headQueue;
+            WeakOrderQueue cursorQueue;
+            WeakOrderQueue prevQueue;
+
+            internal WeakOrderQueue HeadQueue
+            {
+                get { return Volatile.Read(ref this.headQueue); }
+                set { Volatile.Write(ref this.headQueue, value); }
+            }
+
+            internal int Size
+            {
+                get { return this.head; }
+            }
+
+            internal Stack(int maxCapacity, ThreadLocalPool parent, Thread thread)
             {
-                this.Owner = owner;
+                this.maxCapacity = maxCapacity;
+                this.Parent = parent;
                 this.Thread = thread;
+
+                this.elements = new Handle[Math.Min(InitialCapacity, maxCapacity)];
+            }
+
+            internal int IncreaseCapacity(int expectedCapacity)
+            {
+                int newCapacity = this.elements.Length;
+                int maxCapacity = this.maxCapacity;
+                do
+                {
+                    newCapacity <<= 1;
+                }
+                while (newCapacity < expectedCapacity && newCapacity < maxCapacity);
+
+                newCapacity = Math.Min(newCapacity, maxCapacity);
+                if (newCapacity != this.elements.Length)
+                {
+                    Array.Resize(ref this.elements, newCapacity);
+                }
+
+                return newCapacity;
+            }
+
+            internal void Push(Handle item)
+            {
+                Contract.Requires(item != null);
+                if ((item.recycleId | item.lastRecycledId) != 0)
+                {
+                    throw new InvalidOperationException("released already");
+                }
+                item.recycleId = item.lastRecycledId = Thread.CurrentThread.ManagedThreadId;
```

Comment:
pls use the same technique as in netty, in `ThreadLocalPool`:
```
static int IdSource;
int id = Interlocked.Increment(ref idSource);
```
and use the `parent.id` here.